File: 02-Web开发/03-数据库/B+树索引/go/bptree.go
// B+ Tree index - minimal Go implementation.
//
// Mirrors bptree.py and bptree.c in structure; uses M = 4 (max 3 keys/node).
// Inner nodes route via separator keys; leaf nodes hold (key, value) pairs
// and are linked left-to-right via sibling pointers for range scans.
//
// Invariants (per CMU 15-445 Lecture 08 / OpenDSA 7.2 "B-Trees"):
//   * Every leaf at the same depth (perfect balance).
//   * Inner nodes (excluding root) hold >= MIN_KEYS_INNER keys.
//   * Leaves hold >= MIN_KEYS_LEAF keys.
//   * Splits propagate upward; root grows when the root itself splits.
//   * Delete rebalances via borrow-from-sibling, else merges with sibling.
//
// Build: go build .
// Run:   ./bptree
package main

import "fmt"

const (
	M              = 4
	MAX_KEYS       = M - 1
	MIN_KEYS_LEAF  = (M+1)/2 - 1 // 1 for M=4
	MIN_KEYS_INNER = M/2 - 1     // 1 for M=4
)

// Node holds both leaf and inner variants in one struct (tagged via isLeaf).
// Go has no native union, so we keep both fields and use isLeaf to select.
type Node struct {
	isLeaf   bool
	nk       int
	keys     [MAX_KEYS + 1]int    // +1 = split overflow slot
	vals     [MAX_KEYS + 1]string // only for leaves
	next     *Node                // sibling chain, only for leaves
	children [M + 1]*Node         // +1 = split overflow slot, only for inner
}

func leafNew() *Node  { return &Node{isLeaf: true} }
func innerNew() *Node { return &Node{isLeaf: false} }

func freeAll(n *Node) {
	if n == nil {
		return
	}
	if !n.isLeaf {
		for i := 0; i <= n.nk; i++ {
			freeAll(n.children[i])
		}
	}
}

// childIdx returns the smallest i such that keys[i] > k; else nk.
func childIdx(n *Node, k int) int {
	lo, hi := 0, n.nk
	for lo < hi {
		mid := (lo + hi) / 2
		if n.keys[mid] <= k {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}

// keyIdx returns the smallest i such that keys[i] >= k; else len.
func keyIdx(keys []int, k int) int {
	lo, hi := 0, len(keys)
	for lo < hi {
		mid := (lo + hi) / 2
		if keys[mid] < k {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}

// Search performs a point lookup. Returns ("", false) if key is absent.
func Search(root *Node, k int) (string, bool) {
	n := root
	for !n.isLeaf {
		n = n.children[childIdx(n, k)]
	}
	i := keyIdx(n.keys[:n.nk], k)
	if i < n.nk && n.keys[i] == k {
		return n.vals[i], true
	}
	return "", false
}

// RangeQuery walks the leaf sibling chain to collect keys in [lo, hi].
func RangeQuery(root *Node, lo, hi int) [][2]any {
	var out [][2]any
	if root == nil {
		return out
	}
	n := root
	for !n.isLeaf {
		n = n.children[childIdx(n, lo)]
	}
	for n != nil {
		for i := 0; i < n.nk; i++ {
			switch {
			case n.keys[i] < lo:
				continue
			case n.keys[i] > hi:
				return out
			}
			out = append(out, [2]any{n.keys[i], n.vals[i]})
		}
		n = n.next
	}
	return out
}

// splitLeaf splits an over-full leaf in half; returns (copy-up key, new right leaf).
func splitLeaf(L *Node) (int, *Node) {
	mid := L.nk / 2
	R := leafNew()
	R.nk = L.nk - mid
	copy(R.keys[:], L.keys[mid:L.nk])
	copy(R.vals[:], L.vals[mid:L.nk])
	L.nk = mid
	R.next = L.next
	L.next = R
	return R.keys[0], R // copy up
}

// splitInner splits an over-full inner node; middle key is *pushed up*.
// Returns (push-up key, new right inner).
func splitInner(I *Node) (int, *Node) {
	mid := I.nk / 2
	up := I.keys[mid]
	R := innerNew()
	R.nk = I.nk - mid - 1
	for i := 0; i < R.nk; i++ {
		R.keys[i] = I.keys[mid+1+i]
		R.children[i] = I.children[mid+1+i]
	}
	R.children[R.nk] = I.children[I.nk]
	I.nk = mid
	return up, R
}

// leafInsert inserts (k, v) into L (assumed not over-full). Returns true if
// newly inserted, false if k existed (and got value-updated).
func leafInsert(L *Node, k int, v string) bool {
	i := keyIdx(L.keys[:L.nk], k)
	if i < L.nk && L.keys[i] == k {
		L.vals[i] = v
		return false
	}
	for j := L.nk; j > i; j-- {
		L.keys[j] = L.keys[j-1]
		L.vals[j] = L.vals[j-1]
	}
	L.keys[i] = k
	L.vals[i] = v
	L.nk++
	return true
}

// Insert returns the (possibly new) root.
func Insert(root *Node, k int, v string) *Node {
	var path [64]*Node
	var idxPath [64]int
	depth := 0

	n := root
	for !n.isLeaf {
		i := childIdx(n, k)
		path[depth] = n
		idxPath[depth] = i
		depth++
		n = n.children[i]
	}
	if !leafInsert(n, k, v) {
		return root
	}

	var pushedKey int
	var pushedNode *Node
	if n.nk > MAX_KEYS {
		pushedKey, pushedNode = splitLeaf(n)
	}
	for pushedNode != nil && depth > 0 {
		depth--
		par := path[depth]
		idx := idxPath[depth]
		for j := par.nk; j > idx; j-- {
			par.keys[j] = par.keys[j-1]
		}
		par.keys[idx] = pushedKey
		for j := par.nk + 1; j > idx+1; j-- {
			par.children[j] = par.children[j-1]
		}
		par.children[idx+1] = pushedNode
		par.nk++
		if par.nk <= MAX_KEYS {
			pushedNode = nil
			break
		}
		pushedKey, pushedNode = splitInner(par)
	}
	if pushedNode != nil {
		nr := innerNew()
		nr.keys[0] = pushedKey
		nr.children[0] = root
		nr.children[1] = pushedNode
		nr.nk = 1
		return nr
	}
	return root
}

// BulkLoad builds a B+ tree from a SORTED (keys, vals) pair. O(N).
func BulkLoad(keys []int, vals []string) *Node {
	if len(keys) == 0 {
		return leafNew()
	}
	var leaves []*Node
	cur := leafNew()
	for i := 0; i < len(keys); i++ {
		if cur.nk == MAX_KEYS {
			leaves = append(leaves, cur)
			cur = leafNew()
		}
		cur.keys[cur.nk] = keys[i]
		cur.vals[cur.nk] = vals[i]
		cur.nk++
	}
	if cur.nk > 0 {
		leaves = append(leaves, cur)
	}
	for i := 0; i+1 < len(leaves); i++ {
		leaves[i].next = leaves[i+1]
	}
	level := leaves
	for len(level) > 1 {
		var next []*Node
		i := 0
		for i < len(level) {
			par := innerNew()
			par.children[0] = level[i]
			j := i + 1
			for j < len(level) && par.nk < MAX_KEYS {
				par.keys[par.nk] = level[j].keys[0]
				par.children[par.nk+1] = level[j]
				par.nk++
				j++
			}
			next = append(next, par)
			i = j
		}
		level = next
	}
	return level[0]
}

func visualize(n *Node, prefix string, isLast bool) {
	branch := prefix
	if isLast {
		branch += "+-- "
	} else {
		branch += "|-- "
	}
	if n.isLeaf {
		fmt.Printf("%s[leaf ", branch)
		for i := 0; i < n.nk; i++ {
			if i > 0 {
				fmt.Print(", ")
			}
			fmt.Printf("%d:%s", n.keys[i], n.vals[i])
		}
		if n.next != nil {
			fmt.Println("] -> next")
		} else {
			fmt.Println("]")
		}
		return
	}
	fmt.Printf("%s[inner keys=", branch)
	for i := 0; i < n.nk; i++ {
		if i > 0 {
			fmt.Print(",")
		}
		fmt.Printf("%d", n.keys[i])
	}
	fmt.Println("]")
	nextPrefix := prefix
	if isLast {
		nextPrefix += "    "
	} else {
		nextPrefix += "|   "
	}
	for i := 0; i <= n.nk; i++ {
		visualize(n.children[i], nextPrefix, i == n.nk)
	}
}

func countLeaves(n *Node) int {
	if n.isLeaf {
		return 1
	}
	c := 0
	for i := 0; i <= n.nk; i++ {
		c += countLeaves(n.children[i])
	}
	return c
}

File: 02-Web开发/03-数据库/B+树索引/python/bptree.py
"""B+ Tree index - minimal implementation.

In-memory B+ Tree that mirrors the structure used by RDBMS storage engines
(InnoDB clustered index, SQLite, PostgreSQL heap+btree index).

Reference properties for an M-way B+ Tree (per CMU 15-445 Lecture 08,
Andy Pavlo + Jignesh Patel, Fall 2023 / OpenDSA 7.2 "B-Trees", Clifford
Shaffer):

  * Perfectly balanced: every leaf node at the same depth.
  * Inner node other than root: between ceil(M/2)-1 and M-1 keys
    (equivalently between ceil(M/2) and M children).
  * Inner node with k keys has k+1 children.
  * Data lives ONLY in leaf nodes; inner nodes are pure routing.
  * Leaf nodes are linked left-to-right via sibling pointers for range scans.

We fix M = 4 (a.k.a. "order 4") so each node fits at most 3 keys; this is
small enough to print legibly yet exercises split/merge.
"""

from __future__ import annotations

from typing import List, Tuple


M = 4
MAX_KEYS = M - 1
MIN_KEYS_LEAF = (M + 1) // 2 - 1
MIN_KEYS_INNER = M // 2 - 1


class Leaf:
    __slots__ = ("keys", "vals", "next")

    def __init__(self) -> None:
        self.keys: List[int] = []
        self.vals: List[str] = []
        self.next: "Leaf | None" = None


class Inner:
    __slots__ = ("keys", "children")

    def __init__(self) -> None:
        self.keys: List[int] = []
        self.children: List["Node"] = []


Node = Leaf | Inner


# ---------- core utilities ---------------------------------------------------

def is_leaf(n: Node) -> bool:
    return isinstance(n, Leaf)


def find_child_index(n: Inner, k: int) -> int:
    """Return the smallest i s.t. k < n.keys[i]; else len(n.keys)."""
    lo, hi = 0, len(n.keys)
    while lo < hi:
        mid = (lo + hi) // 2
        if n.keys[mid] <= k:
            lo = mid + 1
        else:
            hi = mid
    return lo


def find_key_index(arr: List[int], k: int) -> int:
    """Binary search the smallest i such that arr[i] >= k; else len(arr)."""
    lo, hi = 0, len(arr)
    while lo < hi:
        mid = (lo + hi) // 2
        if arr[mid] < k:
            lo = mid + 1
        else:
            hi = mid
    return lo


# ---------- search / range scan --------------------------------

def search(root: Node, k: int) -> str | None:
    """Point query; search MUST reach a leaf (OpenDSA 7.2.1.1)."""
    n = root
    while not is_leaf(n):
        n = n.children[find_child_index(n, k)]  # type: ignore[assignment]
    i = find_key_index(n.keys, k)
    if i < len(n.keys) and n.keys[i] == k:
        return n.vals[i]
    return None


def range_query(root: Node, lo: int, hi_inclusive: int) -> List[Tuple[int, str]]:
    """Return [(k, v), ...] for lo <= k <= hi_inclusive, in ascending order.

    Descend to the first leaf whose smallest key >= lo, then walk the
    sibling chain emitting keys in range until we pass hi_inclusive.
    """
    if root is None:
        return []
    n = root
    while not is_leaf(n):
        n = n.children[find_child_index(n, lo)]  # type: ignore[assignment]
    out: List[Tuple[int, str]] = []
    while n is not None:
        for k, v in zip(n.keys, n.vals):
            if k < lo:
                continue
            if k > hi_inclusive:
                return out
            out.append((k, v))
        n = n.next  # type: ignore[assignment]
    return out


# ---------- split primitives ------------------------------------------------

def split_leaf(L: Leaf) -> Tuple[int, Leaf]:
    """Split an over-full leaf in half; return (copy-up key, new right leaf)."""
    mid = len(L.keys) // 2
    new = Leaf()
    new.keys = L.keys[mid:]
    new.vals = L.vals[mid:]
    L.keys = L.keys[:mid]
    L.vals = L.vals[:mid]
    new.next = L.next
    L.next = new
    return new.keys[0], new


def split_inner(I: Inner) -> Tuple[int, Inner]:
    """Split an over-full inner node; return (push-up key, new right inner)."""
    mid = len(I.keys) // 2
    up = I.keys[mid]
    new = Inner()
    new.keys = I.keys[mid + 1:]
    new.children = I.children[mid + 1:]
    I.keys = I.keys[:mid]
    I.children = I.children[:mid + 1]
    return up, new


# ---------- insert -----------------------------------------------------------

def insert(root: Node, k: int, v: str) -> Node:
    """Insert (k, v); return (possibly new) root."""
    path: List[Tuple[Inner, int]] = []
    n = root
    while not is_leaf(n):
        i = find_child_index(n, k)  # type: ignore[arg-type]
        path.append((n, i))         # type: ignore[arg-type]
        n = n.children[i]           # type: ignore[assignment]

    L = n  # type: ignore[assignment]
    i = find_key_index(L.keys, k)
    if i < len(L.keys) and L.keys[i] == k:
        L.vals[i] = v
        return root
    L.keys.insert(i, k)
    L.vals.insert(i, v)

    pushed_key: int | None = None
    new_node: Node | None = None
    if len(L.keys) > MAX_KEYS:
        pushed_key, new_node = split_leaf(L)  # type: ignore[arg-type]

    # `parent` is the inner node we are about to splice into; `idx` is its
    # child slot for the just-split child. After the splice, if `parent`
    # itself overflows, we split it and continue propagating.
    parent: Inner | None = None
    while pushed_key is not None and new_node is not None:
        if not path:
            # Root itself split -> new root of height + 1. We splice `parent`
            # (the just-split inner node, now mutated to the left half) and
            # `new_node` (the right half) under a fresh single-key root.
            # Using `L` here would double-link L into two places.
            new_root = Inner()
            new_root.keys = [pushed_key]  # type: ignore[list-item]
            new_root.children = [parent if parent is not None else L, new_node]  # type: ignore[list-item]
            return new_root
        parent, idx = path.pop()
        # Splice: insert separator key at position idx and new_node at idx+1.
        # idx is the original child slot of the just-split child within
        # parent; pushing the separator there shifts the old separator at idx
        # to idx+1 and keeps the keys sorted (the new separator is >= all keys
        # in the just-split left half and < the next sibling's keys).
        parent.keys.insert(idx, pushed_key)  # type: ignore[arg-type]
        parent.children.insert(idx + 1, new_node)  # type: ignore[arg-type]
        if len(parent.keys) <= MAX_KEYS:
            break
        pushed_key, new_node = split_inner(parent)  # type: ignore[arg-type]

    return root


# ---------- delete (implementation lives in bptree_delete.py) ---------------

# Re-exported so `from bptree import delete` keeps working for existing callers
# (demo.py). bptree_delete imports from this module, so the import must stay
# below every definition it needs - which it does, by construction.
from bptree_delete import _delete, delete  # noqa: E402

# ---------- bulk-load --------------------------------------------------------

def bulk_load(pairs: List[Tuple[int, str]]) -> Node:
    """Build a B+ Tree from a SORTED list of (key, value) pairs in O(N).

    Per Wikipedia/CMU: build leaves sequentially, then build inner levels
    bottom-up. This is dramatically faster than N times insert() because
    no path-tracking or split-on-the-way-up is needed.
    """
    if not pairs:
        return Leaf()
    leaves: List[Leaf] = []
    cur = Leaf()
    for k, v in pairs:
        if len(cur.keys) == MAX_KEYS:
            leaves.append(cur)
            cur = Leaf()
        cur.keys.append(k)
        cur.vals.append(v)
    if cur.keys:
        leaves.append(cur)
    for i in range(len(leaves) - 1):
        leaves[i].next = leaves[i + 1]

    level = leaves  # type: ignore[assignment]
    while len(level) > 1:
        new_level: List[Inner] = []
        i = 0
        while i < len(level):
            cur = Inner()
            cur.children = [level[i]]
            j = i + 1
            while j < len(level) and len(cur.keys) < MAX_KEYS:
                cur.keys.append(level[j].keys[0])
                cur.children.append(level[j])
                j += 1
            new_level.append(cur)
            i = j
        level = new_level  # type: ignore[assignment]

    return level[0]  # type: ignore[return-value]


# ---------- tree-shape helpers -----------------------------------------------

def count_leaves(n: Node) -> int:
    if is_leaf(n):
        return 1
    return sum(count_leaves(c) for c in n.children)  # type: ignore[union-attr]